From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: dasd: fix alignment of transport mode recovery TCW
References: bnc#589679
Patch-mainline: Upstream submission made

Symptom:     A recovery TCW fails with CS: 0x20 and schxs: 0x06
Problem:     When the DASD device driver uses the transport mode (High
             Performance FICON) and there is a request that needs error
             recovery, then we create a recovery requests with a fresh recovery
             TCW. Every TCW needs to be aligned on a 64 byte boundary, but for
             the recovery TCWs we fail to do the proper alignment. The first
             recovery request for a device has proper alignment by coincident,
             so most of the times recovery will work. But if there is more then
             one recovery request at the same time, then the recovery will fail
             and result in an I/O error.
Solution:    Do proper alignment for the TCW.

Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/block/dasd_3990_erp.c |    7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

--- a/drivers/s390/block/dasd_3990_erp.c
+++ b/drivers/s390/block/dasd_3990_erp.c
@@ -2270,7 +2270,8 @@ static struct dasd_ccw_req *dasd_3990_er
 
 	if (cqr->cpmode == 1) {
 		cplength = 0;
-		datasize = sizeof(struct tcw) + sizeof(struct tsb);
+		/* TCW needs to be 64 byte aligned, so leave enough room */
+		datasize = 64 + sizeof(struct tcw) + sizeof(struct tsb);
 	} else {
 		cplength = 2;
 		datasize = 0;
@@ -2298,8 +2299,8 @@ static struct dasd_ccw_req *dasd_3990_er
 	if (cqr->cpmode == 1) {
 		/* make a shallow copy of the original tcw but set new tsb */
 		erp->cpmode = 1;
-		erp->cpaddr = erp->data;
-		tcw = erp->data;
+		erp->cpaddr = PTR_ALIGN(erp->data, 64);
+		tcw = erp->cpaddr;
 		tsb = (struct tsb *) &tcw[1];
 		*tcw = *((struct tcw *)cqr->cpaddr);
 		tcw->tsb = (long)tsb;
